aiida/tools/importexport/common/archive.py

```diff
@@ -40,59 +60,135 @@ class Archive(object):
             archive.version
 
     """
+    MANDATORY_DATA_KEYS = {'node_attributes', 'node_extras', 'export_data', 'links_uuid', 'groups_uuid'}
+    MANDATORY_METADATA_KEYS = {'aiida_version', 'export_version', 'all_fields_info', 'unique_identifiers'}
 
-    FILENAME_DATA = 'data.json'
-    FILENAME_METADATA = 'metadata.json'
-
-    def __init__(self, filepath):
+    def __init__(self, filepath, output_filepath=None, overwrite=False, silent=True, sandbox_in_repo=True):
         self._filepath = filepath
+        self._output_filepath = output_filepath
+        self._overwrite = overwrite
+        self._silent = silent
+        self._sandbox_in_repo = sandbox_in_repo
         self._folder = None
         self._unpacked = False
+        self._keep = False
         self._data = None
         self._meta_data = None
+        self._archive_format = None
 
     def __enter__(self):
         """Instantiate a SandboxFolder into which the archive can be lazily unpacked."""
-        self._folder = SandboxFolder()
+        self._folder = SandboxFolder(sandbox_in_repo=self._sandbox_in_repo)
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
-        """Clean the sandbox folder if it was instatiated."""
-        if self.folder:
+        """Clean the sandbox folder if it was instantiated."""
+        if not self.keep and self.folder:
             self.folder.erase()
 
-    @decorator
-    def ensure_within_context(wrapped, instance, args, kwargs):  # pylint: disable=no-self-argument
-        """Decorator to ensure that the instance is called within a context manager."""
-        if instance and not instance.folder:
-            raise InvalidOperation('the Archive class should be used within a context')
+    def _ensure_within_context(self):
```

Comment:
True - it comes from the desire to have a class used only for tests that subclasses `Archive`, but otherwise does not need everything to be in a context. For this I needed the decorator to call an "overridable" class method.